operator/src/main/java/org/bf2/operator/managers/MetricsManager.java

```diff
@@ -0,0 +1,111 @@
+    static final Tag OWNER = Tag.of("owner", "KafkaInstanceMetricsManager");
+
+    @Inject
+    ResourceInformerFactory resourceInformerFactory;
+
+    @Inject
+    MeterRegistry meterRegistry;
+
+    @Inject
+    KubernetesClient kubernetesClient;
+
+    private volatile ResourceInformer<Kafka> kafkaResourceInformer;
+
+    protected synchronized void createInformer() {
+        if (kafkaResourceInformer != null) {
+            return;
+        }
+        kafkaResourceInformer = resourceInformerFactory.create(Kafka.class, filter(kubernetesClient.resources(Kafka.class, KafkaList.class)), new ResourceEventHandler<>() {
+            @Override
+            public void onAdd(Kafka kafka) {
+                createMetrics(kafka);
+            }
+
+            @Override
+            public void onUpdate(Kafka oldObj, Kafka newObj) {
+                createMetrics(newObj);
+            }
+
+            @Override
+            public void onDelete(Kafka obj, boolean deletedFinalStateUnknown) {
+                Search.in(meterRegistry).tags(buildKafkaInstanceTags(obj)).meters().forEach(meterRegistry::remove);
+
+            }
+
+            private void createMetrics(Kafka kafka) {
+                Tags tags = buildKafkaInstanceTags(kafka);
+                meterRegistry.gauge(KAFKA_INSTANCE_BROKERS_DESIRED_COUNT, tags, kafka, this::replicas);
+                meterRegistry.gauge(KAFKA_INSTANCE_PARTITION_LIMIT, tags, kafka, k -> kafkaConfigValue(k, KafkaCluster.MAX_PARTITIONS));
+                meterRegistry.gauge(KAFKA_INSTANCE_MAX_MESSAGE_SIZE_LIMIT, tags, kafka, k -> kafkaConfigValue(k, KafkaCluster.MESSAGE_MAX_BYTES));
+
+                if (kafka.getSpec() != null && kafka.getSpec().getKafka() != null) {
+                    IntStream.range(0, kafka.getSpec().getKafka().getReplicas()).forEach(ordinal -> {
+                        Tags brokerTags = Tags.concat(tags, Tags.of(Tag.of("broker", String.valueOf(ordinal))));
+
+                        if (kafka.getSpec().getKafka().getListeners() != null) {
+                            kafka.getSpec().getKafka().getListeners().stream().filter(l -> "external".equals(l.getName())).forEach(l -> {
+                                Tags listenerTags = Tags.concat(brokerTags, Tags.of(Tag.of("listener", l.getName())));
+                                meterRegistry.gauge(KAFKA_INSTANCE_CONNECTION_LIMIT, listenerTags, l, this::maxConnections);
+                                meterRegistry.gauge(KAFKA_INSTANCE_CONNECTION_CREATION_RATE_LIMIT, listenerTags, l, this::maxConnectionCreationRate);
+                            });
+                        }
+                    });
+                }
+            }
+
+            private Tags buildKafkaInstanceTags(Kafka obj) {
+                ObjectMeta metadata = obj.getMetadata();
+                return Tags.of(Tag.of("namespace", metadata.getNamespace()), Tag.of("name", metadata.getName()), OWNER);
```

Comment:
I guess tags are what Prometheus calls labels? I see a lot of metrics have a `strimzi_io_name` label rather than a `name`. I wonder if there's any reason for that namespacing of label names that we'd need to be aware of?

Reply:
Metric labels that begin `strimzi_io` originate from the pod's labels of pods controlled by strimzi.  

```
oc get pods penguin-kafka-0 --show-labels
NAME              READY   STATUS    RESTARTS   AGE     LABELS
penguin-kafka-0   1/1     Running   0          5h14m   app.kubernetes.io/instance=penguin,app.kubernetes.io/managed-by=strimzi-cluster-operator,app.kubernetes.io/name=kafka,app.kubernetes.io/part-of=strimzi-penguin,controller-revision-hash=penguin-kafka-5d6fd54bcc,ingressType=sharded,managedkafka.bf2.org/kas-multi-zone=true,managedkafka.bf2.org/kas-us-east-1a=true,managedkafka.bf2.org/kas-us-east-1b=true,managedkafka.bf2.org/kas-us-east-1c=true,statefulset.kubernetes.io/pod-name=penguin-kafka-0,strimzi.io/cluster=penguin,strimzi.io/kind=Kafka,strimzi.io/name=penguin-kafka
```

these get mapped into labels by this  [label rewrite rule](https://github.com/bf2fc6cc711aee1a0c2a/observability-resources-mk/blob/6f333220ab8ba5eeccf72373f63900cacad6108c/resources/prometheus/pod_monitors/kafka-metrics.yaml#L20) in the pod monitor into what you see in prometheus.

Should labels be namespaced?  I would argue no, doing so would make joining prometheus metrics produced by different components really hard.  You end up with PromQL full of [`label_replaces()`](https://prometheus.io/docs/prometheus/latest/querying/functions/#label_replace) or record rules that map labels to a common domain.  See what I am doing https://github.com/bf2fc6cc711aee1a0c2a/observability-resources-mk/pull/165 to get to a common set of label names.  Comments/thoughts welcome!